backend/tessend/tessend.go

```diff
@@ -345,27 +391,83 @@ func (t *Tessend) collect(now int64) *Data {
 			Version: version.Semver(),
 			License: wrapper.Value.License,
 		},
-		Metrics: corev2.Metrics{
-			Points: []*corev2.MetricPoint{
-				&corev2.MetricPoint{
-					Name:      "entity_count",
-					Value:     entityCount,
-					Timestamp: now,
-				},
-				&corev2.MetricPoint{
-					Name:      "backend_count",
-					Value:     backendCount,
-					Timestamp: now,
-				},
+	}
+
+	return data
+}
+
+// getTessenMetrics populates the data payload with # backends/entities.
+func (t *Tessend) getTessenMetrics(now int64, data *Data) {
+	var entityCount, backendCount float64
+
+	// collect entity count
+	entities, err := t.store.GetEntities(t.ctx, &store.SelectionPredicate{})
+	if err != nil {
+		logger.WithError(err).Error("unable to retrieve entity count")
+	}
+	if entities != nil {
+		entityCount = float64(len(entities))
+	}
+
+	// collect backend count
+	cluster, err := t.client.Cluster.MemberList(t.ctx)
+	if err != nil {
+		logger.WithError(err).Error("unable to retrieve backend count")
+	}
+	if cluster != nil {
+		backendCount = float64(len(cluster.Members))
+	}
+
+	// populate data payload
+	data.Metrics = corev2.Metrics{
+		Points: []*corev2.MetricPoint{
+			&corev2.MetricPoint{
+				Name:      "entity_count",
+				Value:     entityCount,
+				Timestamp: now,
+			},
+			&corev2.MetricPoint{
+				Name:      "backend_count",
+				Value:     backendCount,
+				Timestamp: now,
 			},
 		},
 	}
+}
 
-	return data
+// getTessenConfigMetrics populates the data payload with an opt-out status event.
+func (t *Tessend) getTessenConfigMetrics(now int64, tessen *corev2.TessenConfig, data *Data) {
+	data.Metrics = corev2.Metrics{
+		Points: []*corev2.MetricPoint{
+			&corev2.MetricPoint{
+				Name:      "tessen_config_update",
+				Value:     1,
+				Timestamp: now,
+				Tags: []*corev2.MetricTag{
+					&corev2.MetricTag{
+						Name:  "opt_out",
+						Value: strconv.FormatBool(tessen.OptOut),
+					},
+				},
+			},
+		},
+	}
 }
 
-// send the data payload to the tessen url
-func (t *Tessend) send(data *Data) (*http.Response, error) {
+// send sends the data payload to the tessen url and retrieves the interval response header.
+func (t *Tessend) send(data *Data) string {
 	b, _ := json.Marshal(data)
-	return http.Post(t.url, "application/json", bytes.NewBuffer(b))
+	resp, err := http.Post(t.url, "application/json", bytes.NewBuffer(b))
+	// TODO(nikki): special case logs on a per error basis
+	if err != nil {
+		logger.WithError(err).Error("tessen phone-home service failed")
+		return ""
+	}
+	if resp.StatusCode >= 400 {
+		body, _ := ioutil.ReadAll(resp.Body)
```

Comment:
We should use `io.LimitReader` here. An attacker could theoretically MITM the tessen service and send an infinite stream of bytes, causing a crash. For instance,

```
body, _ := ioutil.ReadAll(io.LimitReader(resp.Body, 4096))
```